### ISSUE TITLE
Unable to draw image over face detection landmarks via canvas.drawBitmap

### ISSUE THREAD

**Issue by MarianIacob** [firebase/quickstart-android#800]
### Step 1: Describe your environment

  * Android device: Google Pixel
  * Android OS version: 9
  * Google Play Services version: 14.7.99
  * Firebase/Play Services SDK version: exactly as the ones from the latest [sample app](https://github.com/firebase/quickstart-android/tree/master/mlkit):
```
// ML Kit dependencies
    implementation 'com.google.firebase:firebase-ml-vision:19.0.0'
    implementation 'com.google.firebase:firebase-ml-vision-image-label-model:17.0.2'
    implementation 'com.google.firebase:firebase-ml-vision-face-model:17.0.2'
    implementation 'com.google.firebase:firebase-ml-model-interpreter:17.0.1'

classpath 'com.google.gms:google-services:4.2.0'
```

### Step 2: Describe the problem:
I got the latest sample app from [here](https://github.com/firebase/quickstart-android/blob/master/mlkit/README.md) and modified it so that it draws an image over some detected landmarks. This worked fine initially by using _canvas.drawBitmap_ but about one month ago it didn't work anymore. I am not sure if this was related to Android OS upgrade or firebase sdk upgrade.

#### Steps to reproduce:
The things that I changed (from the initial sample code) are:
 - modified constructor of _FaceDetectionProcessor_ from [here](https://github.com/firebase/quickstart-android/blob/master/mlkit/app/src/main/java/com/google/firebase/samples/apps/mlkit/java/facedetection/FaceDetectionProcessor.java#L47) so that it tracks all landmarks:
```
new FirebaseVisionFaceDetectorOptions.Builder()
                        .setLandmarkMode(FirebaseVisionFaceDetectorOptions.ALL_LANDMARKS)
                        .build();
```
 - modified code from _drawLandmarkPosition_ method of _FaceGraphic_ from [here](https://github.com/firebase/quickstart-android/blob/master/mlkit/app/src/main/java/com/google/firebase/samples/apps/mlkit/java/facedetection/FaceGraphic.java#L133) so that it draws a bitmap instead of a circle (bitmap is passed to the FaceGraphic in its constructor):
```
private void drawLandmarkPosition(Canvas canvas, FirebaseVisionFace face, int landmarkID) {
        FirebaseVisionFaceLandmark landmark = face.getLandmark(landmarkID);
        if (landmark != null) {
            FirebaseVisionPoint point = landmark.getPosition();

            Rect dest = new Rect((int) translateX(point.getX() - 5),
                    (int) translateY(point.getY() - 5),
                    (int) translateX(point.getX() + 5),
                    (int) translateY(point.getY() + 5));

            Log.d("MY_LOGS", "Preparing to call drawBitmap with params myBitmap: " + myBitmap + " and dest: " + dest);

            canvas.drawBitmap(myBitmap, null, dest, null);
        }
    }
```
 - started the app
 - selected Java -> LivePreviewActivity
 - switched from "Face Contour" to "Face Detection
 - changed camera to front

#### Observed Results:
The overlayed bitmap does not appear at all, but it is however shown for a split second if I switch the camera (from front to back or back to front).
Should I draw the bitmap in a different way? If so, how? Again, as I mentioned above, this worked fine with no problems until about one month ago. I was not able to figure out what exactly changed in the library or to find a way to fix it.

FYI: the bitmap used is properly decoded and it is a valid image.

Example of log result:
_2019-02-04 15:08:36.395 6298-6298/com.google.firebase.samples.apps.mlkit D/MY_LOGS: Preparing to call drawBitmap with params myBitmap: android.graphics.Bitmap@4ab4cf6 and dest: Rect(915, 80 - 945, 110)_

#### Expected Results:
Expecting the bitmap to be overlayed properly on top of the desired landmark.

#### Relevant Code:
mentioned above

Any help or suggestion is highly appreciated ;)
Thanks in advance!

Cheers,
Marian

**Comment by MarianIacob** [comment]
Hey @samtstern, thanks for the reply!

The overlays in the sample app work fine, but they are using simple shapes like canvas.drawCircle(). Didn't see one that's using drawBitmap.

My first thought was also that this might be Android Canvas API related, but it's still weird that the bitmap appears for a split second when switching cameras. Because of this, I thought that it might be related to clearing and redrawing the overlay. It's also weird that this worked perfectly fine a month ago.
By the way, I tried downgrading to an older mlkit version but this didn't help. Could this be related to Android OS upgrade, from 8.1 to 9? (it's a funky guess, I know :-) )

**Comment by MarianIacob** [comment]
Hi again @samtstern,

I have modified the googly-eyes android vision sample from [here](https://github.com/googlesamples/android-vision/tree/master/visionSamples/googly-eyes), updated all the libs to the latest ones and modified the code so that it uses canvas.drawbitmap and it works fine there. I can fork the code tomorrow if needed. 
This would probably mean that the reported issue might not be canvas api related.

**Comment by samtstern** [comment]
@MarianIacob I think you're right, it's probably something to do with the overlays in this sample app and how we're displaying them.  Remember that this sample app is not meant to show off the best practices for a production Android app, we're just trying to show the capabilities of Firebase in a simple setting.

Since you were able to solve your issue with the android vision sample, is it safe to close this here?  Or do you have a further question about MLKit?

**Comment by MarianIacob** [comment]
Hey @samtstern,

I was able to properly display bitmap overlays with android vision, but as far as I could see, the ears landmarks are not detected there the same way as they are on ml-kit, because of this I am dependent on firebase face detection (our app needs to properly identify the ears).
I know that the sample app is not showing the best practices, but considering that it worked fine and, from my point of view, after an OS and play services upgrade it failed to work properly, I assumed that the sdk itself has some issues. However, considering that the landmark points are properly updated, then it's just a problem with the sample app :-/

I will try to check again today for differences between how the android vision sample displays/invalidates/updates the overlay images and how this is done on the ml-kit sample. 
From my point of view, even if the sample app is not showing the best practices, it should still cover this feature... as it's probably very common (image overlay over live camera face detection).

I will post my findings here (hopefully I will spot something)... and then, if successful, create a pull request ;)

**Comment by MarianIacob** [comment]
Hey @samtstern ,

Good news! Managed to make it work by changing the code:
```
FirebaseVisionFaceLandmark landmark = face.getLandmark(landmarkID);
        if (landmark != null) {
            FirebaseVisionPoint point = landmark.getPosition();

            Rect dest = new Rect((int) translateX(point.getX() - 5),
                    (int) translateY(point.getY() - 5),
                    (int) translateX(point.getX() + 5),
                    (int) translateY(point.getY() + 5));

            canvas.drawBitmap(myBitmap, null, dest, null);
        }
```

... to:
```
FirebaseVisionFaceLandmark landmark = face.getLandmark(landmarkID);
        if (landmark != null) {
            FirebaseVisionPoint point = landmark.getPosition();

            PointF pointPosition =
                        new PointF(translateX(point.getX()), translateY(point.getY()));
            int left = (int) (pointPosition.x - 5);
            int top = (int) (pointPosition.y - 5);
            int right = (int) (pointPosition.x + 5);
            int bottom = (int) (pointPosition.y + 5);

            canvas.drawBitmap(myBitmap, null, new Rect(left, top, right, bottom), null);
        }
```

To be honest I am not really sure why this works, but this was the last thing I could narrow down my ml-kit sample / android-vision sample comparison :) Any idea? Also, I have no idea how the previous code worked one month ago...
I will prepare tomorrow a PR with a bitmap overlay in the sample app. What do you think would be ok to draw? Maybe a clown nose over the NOSE_BASE landmark?

Cheers,
Marian.

**Comment by samtstern** [comment]
@MarianIacob thanks for the follow up!  We will gladly take a PR.  Something simple like a clown nose sounds great.

**Comment by MarianIacob** [comment]
Cool! I will take care of it and get back with a PR ;)

**Comment by MarianIacob** [comment]
Hey @samtstern 

I added a pull request [here](https://github.com/firebase/quickstart-android/pull/808).
Please check it out when you have some time and let me know if some additional adaptations are needed ;)

Cheers,
Marian.

**Comment by samtstern** [comment]
Closed via #808
